### PR TITLE
Secure code from users

### LINKED ISSUE #47: Securing the code sent by users
### Problem

When code is submitted by users, it is unfiltered and can contain malicious elements which cause a serious security threat.

### Fix

When the code is sent by the user a filter should be applied so exploits are minimized. The code should be filtered for:
1. `window.`
2. `document.`
3. `alert`
4. `prompt`

Also, the code should run once completely on the requesters browser as a validation before being sent to the server. This checks for errors in the code and prevents browser-breaking code from being sent to the server.

[Link with example of showing js errors on the web page](https://stackoverflow.com/questions/6604192/showing-console-errors-and-alerts-in-a-div-inside-the-page)

### PR DESCRIPTION
**Pull request by KaspervanM**
### Description

This PR will resolve #47.

When the code is sent by the user a filter will be applied so exploits are minimized. The code will be filtered for:

1. window.
2. document.
3. alert(
4. prompt(
5. confirm(
6. console.
7. eval(
8. import
9. export
10. XMLHttpRequest(
11. fetch(

To be able to use any of the words, use an escape character.

Also, the code will run once completely on the requesters browser as a validation before being sent to the server. This checks for errors in the code and prevents browser-breaking code from being sent to the server.

### CODE REVIEW

**Comment by evanpacini** [comment]
Maybe you can use this regex instead of some difficult checking: 
/window|document|console|alert|prompt|confirm|eval|import|export|XMLHttpRequest|fetch/g

**Comment by KaspervanM** [comment]
> Maybe you can use this regex instead of some difficult checking:
> /window|document|console|alert|prompt|confirm|eval|import|export|XMLHttpRequest|fetch/g

Then how would I know what indexes they can be found at so I can check for the escape character? Or should the escape character be removed entirely?

**Comment by evanpacini** [comment]
> > Maybe you can use this regex instead of some difficult checking:
> > /window|document|console|alert|prompt|confirm|eval|import|export|XMLHttpRequest|fetch/g
> 
> Then how would I know what indexes they can be found at so I can check for the escape character? Or should the escape character be removed entirely?

I thought we were supposed to remove that entirely. At leaast that's what we said in the chat.

**Comment by KaspervanM** [comment]
> You could use .test() for just a boolean return, or .match() for an array of search results, containing the result itself AND the index.
What would you need the index for?

**Comment by evanpacini** [comment]
> > You could use .test() for just a boolean return, or .match() for an array of search results, containing the result itself AND the index.
> 
> What would you need the index for?

I don't think you would need it, but you mentioned index yourself.

File: client/src/views/AddTask.vue [hunk 1]
```diff
@@ -134,54 +149,201 @@ let task: Task = {
...
+    this.task.config = this.config;
+  },
+  watch: {
+    code: function (): void {
+      this.codeErrors = [];
+      for (let i = 0; i < this.codeChecks.length; i++) {
+        if (new RegExp(this.codeChecks[i]).test(this.code)) {
+          this.codeErrors.push("Hey, " + this.codeChecks[i] + " is forbidden!");
+        }
+      }
```
Comment (evanpacini) [inline client/src/views/AddTask.vue:217]: Maybe use .match to simplify the code?
```suggestion
      const regex: RegExp = /window|document|console|alert|prompt|confirm|eval|import|export|XMLHttpRequest|fetch/g;
      const found: Array<string> = regex.match(this.code);
      for (let i = 0; i < found.length; i++) {
        this.codeErrors.push("Hey, " + found[i] + " is forbidden!");
      }
```

File: client/src/views/AddTask.vue [hunk 2]
```diff
@@ -134,54 +149,201 @@ let task: Task = {
...
+        "console",
+        "alert",
+        "prompt",
+        "confirm",
+        "eval",
+        "import",
+        "export",
+        "XMLHttpRequest",
+        "fetch"
+      ],
```
Comment (evanpacini) [inline client/src/views/AddTask.vue:185]: Can be removed if you agree to the following.
```suggestion
```

File: client/src/views/AddTask.vue [hunk 3]
```diff
@@ -134,54 +149,201 @@ let task: Task = {
 };
 export default Vue.extend({
   name: "AddTask",
-  data(): { task: Task; tooltip: string } {
+  data(): {
+    code: string;
+    config: string;
+    codeChecks: Array<string>;
```
Comment (evanpacini) [inline client/src/views/AddTask.vue:155]: Can be removed if you agree to the previous.
```suggestion
```

**Review by evanpacini — APPROVED** [review]
The code works and looks good. Approved for merge!

**Pull request merged**